nautobot/core/apps/__init__.py

```diff
@@ -0,0 +1,249 @@
+
+    Links are specified as Django reverse URL strings.
+    Items are each specified as a list of NavMenuItem instances.
+    """
+
+    permissions = []
+    items = []
+
+    def __init__(self, name, items=None, weight=1000):
+        """Ensure group properties."""
+        self.name = name
+        self.weight = weight
+
+        if items is not None and not isinstance(items, (list, tuple)):
+            raise TypeError("Items must be passed as a tuple or list.")
+        elif not all(isinstance(item, NavMenuItem) for item in items):
+            raise TypeError("All items defined in a group must be an instance of NavMenuItem")
+        self.items = items
+
+
+class NavMenuItem(PermissionsMixin):
+    """
+    This class represents a navigation menu item. This constitutes primary link and its text, but also allows for
+    specifying additional link buttons that appear to the right of the item in the nav menu.
+
+    Links are specified as Django reverse URL strings.
+    Buttons are each specified as a list of NavMenuButton instances.
+    """
+
+    def __init__(self, link, link_text, permissions=None, buttons=None, weight=1000):
+        """Ensure item properties."""
+        super().__init__(permissions)
+        reverse(link)
+        self.link = link
+        self.link_text = link_text
+        self.weight = weight
+
+        if buttons is not None and not isinstance(buttons, (list, tuple)):
+            raise TypeError("Buttons must be passed as a tuple or list.")
+        elif not all(isinstance(button, NavMenuButton) or issubclass(button, NavMenuButton) for button in buttons):
+            raise TypeError("All buttons defined in an item must be an instance or subclass of NavMenuButton")
+        self.buttons = buttons
+
+
+class NavMenuButton(PermissionsMixin):
+    """
+    This class represents a button within a PluginMenuItem. Note that button colors should come from
+    ButtonColorChoices.
+    """
+
+    def __init__(
+        self,
+        link,
+        title,
+        icon_class,
+        button_class=ButtonActionColorChoices.DEFAULT,
+        permissions=None,
+        weight=1000,
+    ):
+        """Ensure button properties."""
+        super().__init__(permissions)
+        reverse(link)
+        self.link = link
+        self.title = title
+        self.icon_class = icon_class
+        self.weight = weight
+
+        if button_class is not None and button_class not in ButtonActionColorChoices.values():
+            raise ValueError("Button color must be a choice within ButtonColorChoices.")
+        self.button_class = button_class
+
+
+class NavMenuAddButton(NavMenuButton):
+    """Add button subclass."""
+
+    def __init__(self, *args, **kwargs):
+        """Ensure button properties."""
+        if "title" not in kwargs:
+            kwargs["title"] = "Add"
+        if "icon_class" not in kwargs:
+            kwargs["icon_class"] = "mdi mdi-plus-thick"
```

Comment:
Thinking about this some more, it would actually be pretty nifty if we at least provided an option for `icon_class` to also be driven from a `ChoiceSet` object, with labels matching the same options as `ButtonActionColorChoices`, something like:

```python
class ButtonActionIconChoices(ChoiceSet):
    """
    Map standard button actions to Material Design Icons classes.
    """

    ADD = "mdi-plus-thick"
    ...
    IMPORT = "mdi-database-import-outline"
    ...
```

(Note also that I'd suggest that the base `"mdi"` class be provided by the `nav_menu.html` template, just as it provides the base `"btn btn-xs"` classes, rather than requiring each button to explicitly include this class.)

This would reduce repetition throughout the UI and make it easy for plugins using this system to provide buttons that have the icons matching the rest of the Nautobot UI **even if we later change which icons we're using for the standard actions**.